### ISSUE TITLE
Question to core/contributors: Does Globalize really need a stash/cache?

### ISSUE THREAD

**Issue by shioyama** [globalize/globalize#567]
This is a question to @globalize/collaborators and also @globalize/core. I'm looking over the code and wondering if we really need this whole stash/cache thing ([`Globalize::ActiveRecord::Attributes`](https://github.com/globalize/globalize/blob/dfc369641c66893ec42de6327284fa095151e04b/lib/globalize/active_record/attributes.rb#L1), [`Globalize::ActiveRecord::Adapter`](https://github.com/globalize/globalize/blob/dfc369641c66893ec42de6327284fa095151e04b/lib/globalize/active_record/adapter.rb#L1), etc.). 

My understanding is that the stash was introduced to track changes on translated attributes and then store them only if there were changes (in `save_translations!`). The cache avoids reads from AR in a similar way.

But looking at the code, it seems that AR should handle all this fine if we just write to the translations directly, I don't see why we need this extra layer which complicates the code tremendously. As a really quick test, I ripped out the stash/cache and replaced it with direct reads/writes to AR, and removed fallbacks/dirty tracking tests which I didn't bother to add support for, and only two specs failed, which seem fixable. Fallbacks and dirty tracking should I believe similarly be handle-able without any stash/cache.

So my question, again: do we need these things? Are they really avoiding hits to the DB? They were introduced way back in globalize2 (maybe earlier? not sure) and Rails2 may not have been as good at handling changes to associations, so maybe the stash was relevant then, but seems to me that now they may be redundant.

@svenfuchs @joshmh I know this is a long long time ago, but if you had any memory of these things it would help. Removing that part of the code would vastly simplify a huge portion of Globalize.

**Comment by svenfuchs** [comment]
@shioyama you're right, it's been a while. I don't remember exactly, but yeah, I do believe I found ActiveRecord not reliable at these things at that time. 

I guess the change tracking could be removed as you suggest. I'm not sure about the caching though. Does ActiveRecord now use caching by default? If used outside of Rails it used to be so that caching was not enabled unless you'd manually opt in.

**Comment by shioyama** [comment]
@svenfuchs Wow thanks for the quick reply! Glad to hear that my hunch was right. I agree the stash seems easier to rip out than the cache, and I'm not totally sure that AR will handle things cleanly but the tests at least don't seem to check for whatever magic those parts of the code are performing.

> Does ActiveRecord now use caching by default?

Actually I'm not sure about that, but worth checking. If it does not, and Globalize fills in for that, it should be in the tests at least. But I have a hunch the cache may not really be doing much as a layer.

**Comment by shioyama** [comment]
So, having looked into this, I realized it doesn't look like Globalize even has a stash anymore. [This commit](https://github.com/globalize/globalize/commit/a2214fc161748b9f0eb5a7fae5e2727ae4481c5b) from way back in 2011 removed it, then [here](https://github.com/globalize/globalize/commit/6c957618bc31f14762ec22015956c49107e40865) again in 2011 it was added back in a different form, but the `translations` accessor that `Adapter` was using is no longer actually used (I just removed it in #576 now).

So unless I'm misunderstanding something, the comment [here](https://github.com/ballistiq/globalize/blob/4b85e61cb1cb501e471f46ea2886ad718a56a91a/lib/globalize/active_record/adapter.rb#L4) should be removed to avoid confusion.

If I'm not misunderstanding, then Globalize only has a stash, and no cache. The stash is only updated on writes, not reads, so every time a translation is fetched (unless we just wrote it) we're loading all translations and looping through them to find the one we want. Which seems totally wrong...

So actually I've done a 180 and my updated question is: Why does Globalize not have a cache? It seems like this must be hitting performance.
